### DISCUSSION TITLE
VReplication: ‘VStreamRows’ and ‘VStream’ are pull based or push based?

### DISCUSSION THREAD

**Discussion by earayu** [wesql/wescale#307] [Q&A] [answered]
‘VStreamRows’ and ‘VStream’ are pull based or push based?

**Comment by gerayking** [comment]
pull.
In Vreplication, the VstreamRows is invoked within vcopier.copyTable. Hence, I believe the table migration in onlineDDL should employ the 'pull' approach.

**Reply by earayu** [reply, marked as answer]
```
  // VStream streams vreplication events.
  rpc VStream(binlogdata.VStreamRequest) returns (stream binlogdata.VStreamResponse) {};

  // VStreamRows streams rows from the specified starting point.
  rpc VStreamRows(binlogdata.VStreamRowsRequest) returns (stream binlogdata.VStreamRowsResponse) {};
```

But based on the above gRPC definition, it appears that they are using a producer push-based approach.
